### ISSUE TITLE
Hydra usage feedback

### ISSUE THREAD

**Issue by omry** [EulerFaria/hydra-pipeline-ex#2]
Hi, thanks for putting this together!
I am happy to this this, please consider writing Medium post about your package!

Taking a quick glance at your repo to provide some basic feedback.

Do not convert OmegaConf to a container unless you have a very good reason.
https://github.com/EulerFaria/hydra-pipeline-ex/blob/6b311b2f61137c1674d3098923ff94d721a7bd31/main.py#L16

Because you convert, the rest of your code is much less elegant:
For example:

https://github.com/EulerFaria/hydra-pipeline-ex/blob/6b311b2f61137c1674d3098923ff94d721a7bd31/train.py#L18

Can become:
```python
 if cfg.scaling.type == 'std': 
```
You are loosing some other features as well. Please don't do that (definitely not as an example).

https://github.com/EulerFaria/hydra-pipeline-ex/blob/a9f6d71424d6f21edb02fd57d4579fd3cb57d8be/conf/model/lr.yaml#L2
No need to quote strings in YAML (unless they contain some special characters. even space is okay usually!)

# HPO
https://github.com/EulerFaria/hydra-pipeline-ex/blob/6b311b2f61137c1674d3098923ff94d721a7bd31/train.py#L45

Hydra 1.0 (the next version) will have two plugins for Hyper parameter optimization you might find useful.
The [Ax plugin](https://hydra.cc/docs/next/plugins/ax_sweeper) is already in master, and the Nevergrad plugin has an [advanced PR](https://github.com/facebookresearch/hydra/pull/430) and will be added soon.

By the way, please join the Hydra chat if you have any questions.

**Comment by sleao** [comment]
Even tho it runs, I'm still not 100% happy with how I handled the line below. In this case, I was running into errors when trying to pass the ListConfig as an argument to GridSearchCV, since it won't recognize the ListConfig as a List (or a nparray). So I converted it using the to_container function. Does ListConfig has a method to pass it as a default List? Is there a better way to do what I did?

`model = OmegaConf.to_container(cfg.model)`

**Comment by omry** [comment]
If you have an API that does an instancecheck for a list you can convert like you did when you pass to it.

Convert just the object you need to convert, and on the callsite. not the whole thing in advance.
Also use resolve=True to convert any interpolations to their final value.
```
model = OmegaConf.to_container(cfg.model, resolve=True)
```

**Comment by EulerFaria** [comment]
Thanks for the help and tips @sleao. 

My next steps will be adding at least one more model and then investigate [Ax plugin](https://hydra.cc/docs/next/plugins/ax_sweeper/) as @omry suggested.  

Once I conclude that I will let you know about the results and close this issue.  

I appreciate all the feedback guys.

**Comment by omry** [comment]
The Ax plugin is new in the master branch. Hydra 1.0 is not officially released yet but it would be awesome if you can give it a try from master (as I suggested already, feel free to join the [chat](https://hydra-framework.zulipchat.com/#) if you run into issues or have questions).

That version comes with many big improvements, the most significant of which is Structured configs.
You can check the new tutorial [here](https://hydra.cc/docs/next/tutorials/intro) (The tutorial is not as polished as the basic tutorial yet).

Side note: I recommend that you add some example of [interpolation](https://omegaconf.readthedocs.io/en/1.4_branch/usage.html#variable-interpolation) to your tutorial. this is a very important feature of OmegaConf (and Hydra naturally).

**Comment by EulerFaria** [comment]
Hey guys, 

One more time, thank you for the help and tips.  

I tested the Ax plugin and I found interesting the approach, however, I noticed it will limit the use of other hyperparameter optimization options such as RandonSearchCV, BayesianOptimization, etc... Moreover, the fact that you have to pass the range of the hype parameters in the command line makes it more difficult to test different models.  

So, I will close this issue and keep monitoring the releases of hydra. If I find a better way I surely share and let y'all know.

**Comment by omry** [comment]
The plugin does support defining ranges and type for variables in the config file.
As for other optimization methods, Ax does support a few - however I don't think we tested anything other than the default.

cc @shagunsodhani

**Comment by omry** [comment]
And by the way, in addition to the Ax plugin ([docs](https://hydra.cc/docs/next/plugins/ax_sweeper/) in case you missed them), we landed the [Nevergrad plugin](https://hydra.cc/docs/next/plugins/nevergrad_sweeper/) to master.
Nevergrad is another Hyper parameter optimization package.

FYI: The interface of both plugins will be changed in the coming weeks or so to be compatible, so you will be able to more easily alternate between them.

**Comment by EulerFaria** [comment]
That's nice! If we could apply it in a multimodel scenario, taking advantage of multirun to create new possibilities of experiments with Pipeline it would be awesome

**Comment by omry** [comment]
Unless I misunderstand you, this is possible in principle with both packages.
You need to specify a configuration choice (a,b,c).
for example,
```
--multirun model=a,b,c lr=0:1
```

This has not yet been stress tested and may require some toolkit specific tweaks to get it to work well, but in principle it's there.

**Comment by EulerFaria** [comment]
In the case you pointed out, the lr (learning rate) must be a parameter for all models a,b,c. In the example proposed, we have two models (e.g. LogisticRegression and XGBoost) with particular parameters for each one disposed on the config files.  

` --multirun scaling=std,minmax model=logistic,xgboost`

It is a bit different than trying out with different ANN architectures which usually present the same group of parameters to test (e.g. lr, weigh_decay, etc.. )  

The idea with this example besides providing reproducibility of the experiments is building a framework that could automatically generate a set of different test cases using Pipeline filled with different pre-determined transformations and a selected group of models (e.g. SVM, XGBoost, LGBM, etc.. )

**Comment by shagunsodhani** [comment]
> The plugin does support defining ranges and type for variables in the config file.
> As for other optimization methods, Ax does support a few - however I don't think we tested anything other than the default.
> 
> cc @shagunsodhani

Hey @EulerFaria Ax supports Bayesian optimization and bandit optimization. For now, we rely on Ax to select the best strategy though we can add an option for the user to specify :)

**Comment by omry** [comment]
@EulerFaria, I am not claiming that the Ax plugin can do anything you might think of :)
This is another tool you gain access to by using Hydra. use it if it makes sense for you.
